File: solutions/clojure/bird-watcher/1/src/bird_watcher.clj
(ns bird-watcher)

;; return a vector of bird counts
(def last-week
  [0 2 5 3 7 8 4]
  )

;; get the last element from the vector
(defn today [birds]
  (last birds)
  )

;; increment today's count
(defn inc-bird [birds]
  (assoc birds 6 (+ 1 (today birds))))

;; was there a day with no visiting birds?
(defn day-without-birds? [birds]
  (not-every? pos? birds)
  )

;; calculate the number of visiting birds for the first number of days
(defn n-days-count [birds n]
  (reduce + (take n birds))
  )

;; calculate the number of busy days, where busy days are 5 or more birds
(defn busy-days [birds]
  (count (filter #(> % 4) birds))
  )

;; do the counts alternate between one and zero birds visiting?
;; test both alternating patterns
(defn odd-week? [birds]
  (or
    (= birds [1,0, 1, 0, 1, 0, 1])
    (= birds [0, 1, 0, 1, 0, 1, 0]))
  )

File: solutions/go/bird-watcher/1/bird_watcher.go
package birdwatcher

// TotalBirdCount return the total bird count by summing
// the individual day's counts.
func TotalBirdCount(birdsPerDay []int) int {
	var total int
	for i := 0; i < len(birdsPerDay); i++ {
		total = total + birdsPerDay[i]
	}
	return total
}

// BirdsInWeek returns the total bird count by summing
// only the items belonging to the given week.
func BirdsInWeek(birdsPerDay []int, week int) int {
	var total int
	// add seven days of birds, index offset by the week number
	for i := 0; i < 7; i++ {
		total = total + birdsPerDay[((week-1)*7)+i]
	}
	return total
}

// FixBirdCountLog returns the bird counts after correcting
// the bird counts for alternate days.
func FixBirdCountLog(birdsPerDay []int) []int {
	for i := 0; i < len(birdsPerDay); i++ {
		if i%2 == 0 {
			// odd day, count the extra bird
			birdsPerDay[i] = birdsPerDay[i] + 1
		}
	}
	return birdsPerDay
}
